contracts/interfaces/oracles/StandardConstraintsOracle.sol

```diff
@@ -0,0 +1,139 @@
+    error InvalidGreaterThanConstraintValue();
+    error InvalidLowerThanConstraintValue();
+
+    event Finalize(uint256 result);
+
+    function __ConstrainedOracle_init(
+        address _kpiToken,
+        uint256 _templateId,
+        uint128 _templateVersion,
+        Constraint _constraint,
+        uint256 _value0,
+        uint256 _value1
+    ) internal onlyInitializing {
+        if (_kpiToken == address(0)) revert ZeroAddressKPIToken();
+        if (_templateId == 0) revert InvalidTemplateId();
+        if (_templateVersion == 0) revert InvalidTemplateVersion();
+
+        if (_value0 == INVALID_ANSWER || _value1 == INVALID_ANSWER) revert InvalidValues();
+
+        if (_constraint == Constraint.Between || _constraint == Constraint.Range) {
+            if (_value1 <= _value0) {
+                revert InvalidBetweenConstraintValues();
+            }
+        } else if (_constraint == Constraint.GreaterThan) {
+            if (_value0 >= INVALID_ANSWER - 1) {
+                revert InvalidGreaterThanConstraintValue();
+            }
+        } else if (_constraint == Constraint.LowerThan) {
+            if (_value0 == 0) {
+                revert InvalidLowerThanConstraintValue();
+            }
+        } else if (_constraint == Constraint.Equal) {
+            // no extra validation is required for the "equal" constraint as the only invalid
+            // value would be the invalid answer one, which is covered above
+        } else {
+            // we should never arrive here
+            revert InvalidConstraint();
+        }
+
+        kpiToken = _kpiToken;
+        oraclesManager = msg.sender;
+        templateId = _templateId;
+        templateVersion = _templateVersion;
+        constraint = _constraint;
+        value0 = _value0;
+        value1 = _value1;
+    }
+
+    function _finalize(uint256 _result) internal {
+        if (finalized) revert Forbidden();
+        finalized = true;
+        uint256 _completionPercentage = _resultToCompletionPercentage(_result);
+        IKPIToken(kpiToken).finalize(_completionPercentage);
+        emit Finalize(_result);
+    }
+
+    function _resultToCompletionPercentage(uint256 _result) internal view returns (uint256) {
+        if (_result == INVALID_ANSWER) {
+            return INVALID_ANSWER;
+        }
+
+        Constraint _constraint = constraint;
+        if (_constraint == Constraint.Range) {
+            uint256 _value0 = value0;
+            uint256 _value1 = value1;
+            if (_result <= _value0) return 0;
+            if (_result >= _value1) {
+                return UNIT;
+            } else {
+                uint256 _numerator = (_result - _value0 * UNIT) << MULTIPLIER;
+                uint256 _denominator = _value1 - _value0;
+                return (_numerator / _denominator) >> MULTIPLIER;
+            }
+        } else if (_constraint == Constraint.Between) {
+            return _result >= value0 && _result <= value0 ? UNIT : 0;
```

Comment:
Here we should check that the `_result` is less than `value1` and not `value0`, no?

Reply:
Absolutely, should be fixed on the new push, where I've also added tests.